From f7bef9882d0fba25d1b63766ff0fe791c830e7c7 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 7 Jul 2023 11:46:12 +0200
Subject: [PATCH] WLANRTSYS-56718: IOP fixes implementation

- Original issue:
    WLANRTSYS-4351 fixed: wpa_supplicant: malformed packet sent by wpa_s
    disable sending L2 wnm keep-alive frames.
    Skip sending keep-alive to AP till frame format will be defined
---
 src/ap/ieee802_11_vht.c      | 6 ++++++
 src/common/ieee802_11_defs.h | 1 +
 wpa_supplicant/events.c      | 8 +-------
 3 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/src/ap/ieee802_11_vht.c b/src/ap/ieee802_11_vht.c
index 607209529..e7eafc965 100644
--- a/src/ap/ieee802_11_vht.c
+++ b/src/ap/ieee802_11_vht.c
@@ -65,6 +65,12 @@ u8 * hostapd_eid_vht_capabilities(struct hostapd_data *hapd, u8 *eid, u32 nsts)
 	/* Supported MCS set comes from hw */
 	os_memcpy(&cap->vht_supported_mcs_set, mode->vht_mcs_set, 8);
 
+	if (CONF_OPER_CHWIDTH_160MHZ == hostapd_get_oper_chwidth(hapd->iconf)) {
+		u16 tx_highest = le_to_host16(cap->vht_supported_mcs_set.tx_highest);
+		tx_highest |= VHT_CAP_EXT_NSS_BW_CAPABLE;
+		cap->vht_supported_mcs_set.tx_highest = host_to_le16(tx_highest);
+	}
+
 	pos += sizeof(*cap);
 
 	return pos;
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 62fb8b0eb..6e4cb27c7 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -1424,6 +1424,7 @@ struct auth_fail_list {
 #define VHT_OPMODE_NOTIF_RX_NSS_SHIFT		    4
 
 #define VHT_RX_NSS_MAX_STREAMS			    8
+#define VHT_CAP_EXT_NSS_BW_CAPABLE		    ((u16) BIT(13))
 
 #define VHT_OPMODE_CHANNEL_40MHZ		    ((u8) BIT(0))
 #define VHT_OPMODE_CHANNEL_80MHZ		    ((u8) BIT(1))
diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 9ecbbeb50..90a2f9c27 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -2835,14 +2835,8 @@ static void wnm_bss_keep_alive(void *eloop_ctx, void *sock_ctx)
 		return;
 
 	if (!wpa_s->no_keep_alive) {
-		wpa_printf(MSG_DEBUG, "WNM: Send keep-alive to AP " MACSTR,
+		wpa_printf(MSG_DEBUG, "WNM: Skip sending keep-alive to AP " MACSTR,
 			   MAC2STR(wpa_s->bssid));
-		/* TODO: could skip this if normal data traffic has been sent */
-		/* TODO: Consider using some more appropriate data frame for
-		 * this */
-		if (wpa_s->l2)
-			l2_packet_send(wpa_s->l2, wpa_s->bssid, 0x0800,
-				       (u8 *) "", 0);
 	}
 
 #ifdef CONFIG_SME
-- 
2.43.0

